astacus/coordinator/plugins/clickhouse/async_object_storage.py

```diff
@@ -0,0 +1,59 @@
+"""
+Copyright (c) 2023 Aiven Ltd
+See LICENSE for details
+"""
+from abc import ABC, abstractmethod
+from pathlib import Path
+from rohmu.errors import FileNotFoundFromStorageError
+from starlette.concurrency import run_in_threadpool
+
+import dataclasses
+import datetime
+import logging
+import rohmu
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(frozen=True)
+class ObjectStorageItem:
+    key: Path
+    last_modified: datetime.datetime
+
+
+class AsyncObjectStorage(ABC):
+    @abstractmethod
+    async def list_items(self) -> list[ObjectStorageItem]:
+        ...
+
+    @abstractmethod
+    async def delete_item(self, key: Path) -> None:
+        ...
+
+
+@dataclasses.dataclass(frozen=True)
+class RohmuAsyncObjectStorage(AsyncObjectStorage):
+    storage: rohmu.BaseTransfer
+
+    async def list_items(self) -> list[ObjectStorageItem]:
+        items = await run_in_threadpool(self.storage.list_iter, key="", with_metadata=True, deep=True)
```

Comment:
Concurrent list and delete are not really a problem. Or more accurately, they are an unavoidable reality, because even of we had a threading or asyncio lock there, the storage is far away in a remote server. There can be multiple instances of `RohmuAsyncObjectStorage`, or multiple processes touching the same bucket, etc. 

However, yes, at least for S3, rohmu keeps a botocore session in its state, which is not thread safe, so the `storage` object itself should be locked to avoid corrupting it :+1:

Reply:
Yup, concurrent list and delete are ok with multiple storage instances; it's using the same ```storage``` instance concurrently which is not fine (resulted in coredump with GCP as well if I remember my rohmu tests).